### DISCUSSION TITLE
How can I "publish" my extension?

### DISCUSSION THREAD

**Discussion by AignerGames** [eclipse-langium/langium#1779] [Q&A] [answered]
I followed the langium tutorials and I have a working language which works when I debug the extension and custom a CLI command which I can call from my development project. Now my next question is, how can I actually "publish" and use these, instead of running it from my VS Code project? 

My goal:
- Create and install my language as an extension for my own VS Code (not a "public" extension, just a local installer for me)
- With the extension installed, my VS Code should use the langium parser for my file extension
- With the extension install, my VS Code should somehow be able to execute my custom cli command

I assume all of this is already supported and Im just missing the final puzzle piece to "build" the extension?

**Comment by msujew** [comment, marked as answer]
Hey @AignerGames,

for most VS Code related questions you can refer to their [docs (on publishing extensions)](https://code.visualstudio.com/api/working-with-extensions/publishing-extension). It explains the process of packaging and publishing extensions in detail.

Before publishing, I would recommend to manually install the created `.vsix` file once into your VS Code instance to ensure that everything works as expected.

> With the extension install, my VS Code should somehow be able to execute my custom cli command

How do you execute the CLI command? Is it bundled together with the extension code or does it run as a separate process (I.e. as a real CLI tool)?

**Reply by AignerGames** [reply]
Thank you for the answer, I will check their documentation. 

Regarding the CLI command, I use the CLI code that came with the langium project and adjusted the code inside the "generate" command (which would generate some JavaScript code by default). Currently I run it in my VSCode project by executing node ./bin/cli args inside the VS Code project terminal. If possible I would like to somehow add this bundled inside the extension code, so that I can run the same command from inside VS Code after the extension was installed, is that possible?

**Reply by msujew** [reply]
I see, yes, that's what I was getting at. I would recommend you to [create a command in your vscode extension](https://code.visualstudio.com/api/extension-guides/command#creating-new-commands) that calls the CLI code directly. The bundler should take care of bundling everything together so that it's available for consumers of your extension.

**Reply by AignerGames** [reply]
Thank you for the information and the link, I refactored my command code a bit and now I can call it from a vscode extension command

**Reply by AignerGames** [reply]
Update, I managed to create a working vsix file with my extension, here are a few notes:

1, vsce package always complained about a missing entry point inside a main.cjs.js file, which makes no sense because the entrypoint is defined as just main.cjs, it looks like this was a bug in a older version? I upgraded the vsce version (with: npm install --force -g @vscode/vsce) and then it didn't show this error anymore

2, But then it complained about a different error, it said the the .vscodeignore can't be used if the package.json contains a files section and the default langium project contains both, so I wasn't sure what to do and just deleted the .vscodeignore and hoped for the best

3, The .vsix was finally created and after installing it in my local VS Code the language and the command worked, but I noticed that the syntax lost it's color. I don't know how the syntax color works, but I assumed that it was related to the files inside the syntax folder, which was now ignored. 

4, I adjusted the files section inside the package.json to include "syntaxes" and built the vsxi again... and it finally works, the language, the colors and the custom command
